### PR TITLE
sync: overlap check is now filter-sensitive

### PR DESCRIPTION
**Pull request by Schlossgeist**
#### What is the purpose of this change?

Previously, the overlap check was based on simple prefix checks of the source and destination paths. Now it actually checks whether the destination is excluded via any filter rule or a "--exclude-if-present"-file.

#### Was the change discussed in an issue or in the forum before?

[Forum Post](https://forum.rclone.org/t/excluding-overlapping-destination/29591)

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by Schlossgeist** [comment]
I've noticed some initially unwanted (but maybe reasonable) behavior. Because of the way I implemented the recursive checks, the sync command is only allowed to be executed when the destination directory already exists (in contrast to sync commands under normal circumstances where the destination directory is created automatically).
Although I think that the ability to sync to subdirectories which are exculded from the sync is inherently useful, I also think that this is kind of an edge-case usecase which should required some additional action from the user (i. e. creating the destination directory). It also gives beginner users the impression that it is generally not ok to try and sync or move directories in this way.

What are your thoughts on this? @ncw

**Comment by ncw** [comment]
> I've noticed some initially unwanted (but maybe reasonable) behavior. Because of the way I implemented the recursive checks, the sync command is only allowed to be executed when the destination directory already exists

This is because you are listing the directory I think.

It should be possible to do these checks without actually calling `List` shouldn't it? I'm concerned about calling `List` since it is potentially an expensive operation (think a bucket with 1 million entries) so we don't want to do it twice.

> (in contrast to sync commands under normal circumstances where the destination directory is created automatically). Although I think that the ability to sync to subdirectories which are exculded from the sync is inherently useful, I also think that this is kind of an edge-case usecase which should required some additional action from the user (i. e. creating the destination directory). It also gives beginner users the impression that it is generally not ok to try and sync or move directories in this way.

I'm not sure I understood the problem there! Can you give some examples please?

I will give the code a quick first review now.

File: fs/operations/operations.go [hunk 1]
```diff
@@ -796,6 +796,53 @@ func Overlapping(fdst, fsrc fs.Info) bool {
 	return strings.HasPrefix(fdstRoot, fsrcRoot) || strings.HasPrefix(fsrcRoot, fdstRoot)
 }
 
+// OverlappingFilterCheck returns true if fdst and fsrc point to the same
+// underlying Fs and they overlap without fdst being excluded by any filter rule.
+func OverlappingFilterCheck(ctx context.Context, fdst fs.Fs, fsrc fs.Fs) bool {
```
Comment (ncw) [inline fs/operations/operations.go:801]: This needs testing in `operations_test.go`

File: fs/operations/operations.go [hunk 2]
```diff
@@ -796,6 +796,53 @@ func Overlapping(fdst, fsrc fs.Info) bool {
...
+	fsrcRoot := fixRoot(fsrc)
+	if strings.HasPrefix(fdstRoot, fsrcRoot) {
+		fdstRelative := strings.ReplaceAll(fdstRoot, fsrcRoot, "")
+		return FilterCheckR(ctx, fdstRelative, "", fsrc)
+	}
+	return strings.HasPrefix(fsrcRoot, fdstRoot)
+}
+
+// FilterCheckR checks if fdst would be included in the sync
+func FilterCheckR(ctx context.Context, fdstRelative string, inputPath string, fsrc fs.Fs) bool {
```
Comment (ncw) [inline fs/operations/operations.go:815]: This needs testing in `operations_test.go` also

File: fs/operations/operations.go [hunk 3]
```diff
@@ -796,6 +796,53 @@ func Overlapping(fdst, fsrc fs.Info) bool {
...
+	include := true
+	fi := filter.GetConfig(ctx)
+	includeDirectory := fi.IncludeDirectory(ctx, fsrc)
+	entries, _ := fsrc.List(ctx, inputPath)
```
Comment (ncw) [inline fs/operations/operations.go:819]: Really don't like calling `List` here. We should be able to do this just by examining the filter and paths shouldn't we?

File: fs/sync/sync_test.go [hunk 4]
```diff
@@ -1442,6 +1442,38 @@ func TestSyncOverlap(t *testing.T) {
...
+	defer r.Finalise()
+
+	fi, err := filter.NewFilter(nil)
+	require.NoError(t, err)
+	require.NoError(t, fi.Add(false, "/rclone-sync-test/"))
+	ctx = filter.ReplaceConfig(ctx, fi)
+
+	subRemoteName := r.FremoteName + "/rclone-sync-test"
+	FremoteSync, err := fs.NewFs(ctx, subRemoteName)
+	FremoteSync.Mkdir(ctx, "")
```
Comment (ncw) [inline fs/sync/sync_test.go:1458]: The error for this needs to be checked. 

    require.NoError(t, FremoteSync.Mkdir(ctx, ""))

is fine

**Review by ncw — COMMENTED** [review]
A quick initial review :-)

Thank you

File: fs/operations/operations.go [hunk 5]
```diff
@@ -796,6 +796,57 @@ func Overlapping(fdst, fsrc fs.Info) bool {
...
+// OverlappingFilterCheck returns true if fdst and fsrc point to the same
+// underlying Fs and they overlap without fdst being excluded by any filter rule.
+func OverlappingFilterCheck(ctx context.Context, fdst fs.Fs, fsrc fs.Fs) bool {
+	if !SameConfig(fdst, fsrc) {
+		return false
+	}
+	fdstRoot := fixRoot(fdst)
+	fsrcRoot := fixRoot(fsrc)
+	if strings.HasPrefix(fdstRoot, fsrcRoot) {
+		fdstRelative := strings.ReplaceAll(fdstRoot, fsrcRoot, "")
```
Comment (ncw) [inline fs/operations/operations.go:808]: Probably better written something like this (`ReplaceAll` could replace stuff it isn't supposed to).

    fdstRelative := fdstRoot[len(fsrcRoot):]

File: fs/operations/operations.go [hunk 6]
```diff
@@ -796,6 +796,57 @@ func Overlapping(fdst, fsrc fs.Info) bool {
...
+			if currentPath == fdstRelative {
+				return true
+			}
+			pos++
+			include = FilterCheckR(ctx, fdstRelative, pos, fsrc)
+		}
+	}
+
+	if !include {
+		err := fsrc.Mkdir(ctx, fdstRelative)
```
Comment (ncw) [inline fs/operations/operations.go:841]: It shouldn't be necessary to run `Mkdir` here - the sync will do it for you.

File: fs/operations/operations.go [hunk 7]
```diff
@@ -796,6 +796,57 @@ func Overlapping(fdst, fsrc fs.Info) bool {
...
+func FilterCheckR(ctx context.Context, fdstRelative string, pos int, fsrc fs.Fs) bool {
+	include := true
+	fi := filter.GetConfig(ctx)
+	includeDirectory := fi.IncludeDirectory(ctx, fsrc)
+	dirs := strings.SplitAfterN(fdstRelative, "/", pos+2)
+	newPath := ""
+	for i := 0; i <= pos; i++ {
+		newPath += dirs[i]
+	}
+	dir := fs.NewDir(newPath, time.Now())
```
Comment (ncw) [inline fs/operations/operations.go:824]: I don't think this `dir` gets used anywhere does it? How about `currentPath := newPath`?

**Review by ncw — COMMENTED** [review]
I think this looks much better now.

I don't think you need to call `Mkdir` either.

Can you write a table driven test for `FilterCheckR` testing all the corner cases - it is quite complicated that function and I'd like to be sure it is correct!

A test for `OverlappingFilterCheck` would be nice too.

Thank you :-)

**Comment by Schlossgeist** [comment]
I used the existing `TestOverlapping` for _inspiration_. I only wrote a test for `OverlappingFilterCheck` though. Can you explain why there should be a separate test for `FilterCheckR`? This function is only called from `OverlappingFilterCheck` with almost the same arguments every time. It feels a bit redundant to test the same function in `sync_test.go` and twice in `operations_test.go`.

**Comment by ncw** [comment]
> I used the existing `TestOverlapping` for _inspiration_. I only wrote a test for `OverlappingFilterCheck` though. Can you explain why there should be a separate test for `FilterCheckR`? This function is only called from `OverlappingFilterCheck` with almost the same arguments every time. It feels a bit redundant to test the same function in `sync_test.go` and twice in `operations_test.go`.

It might be that `FilterCheckR` shouldn't be an exported function, if so rename it `filterCheckR` and then you don't need to make tests for it :-)

**Comment by Schlossgeist** [comment]
@ncw Is this PR ready for merging or should I change anything else?

**Comment by Schlossgeist** [comment]
No problem. Thank you for investing your time and providing this great piece of software for free.
